content/documentation/pages/3-stream-developer-guides/3-programming-models/0-index.md

```diff
@@ -5,6 +5,196 @@ description: 'Programming models'
 summary: true
 ---
 
-# Advanced Stream Processing
+# Programming Model for Streaming Applications
 
-Discuss at a high level what is covered in the following guides.
+Spring Cloud Stream provides the flexibility to build the streaming applications using different programming models.
+
+- _Message Channels_
+- _Functional_
+- _Kafka Streams_
+- _Reactive_
+
+In the following section, we will review how a business logic can be built with different programming models.
+
+## Demonstration
+
+To highlight the use of programming with a concrete example, let's think of a scenario where we are receiving data from an HTTP endpoint. Once when the data is available, suppose we would want to transform the payload by adding prefix and suffixes. Finally, we would want to verify the transformed data.
+
+### Download Applications
+
+To demonstrate the before mentioned use-case, we will start by downloading two out-of-the-box applications.
+
+- [HTTP Source](https://github.com/spring-cloud-stream-app-starters/http)
+
+```bash
+wget http://repo.spring.io/release/org/springframework/cloud/stream/app/http-source-kafka/2.1.0.RELEASE/http-source-kafka-2.1.0.RELEASE.jar
+```
+
+- [Log Sink](https://github.com/spring-cloud-stream-app-starters/log)
+
+```bash
+wget http://repo.spring.io/release/org/springframework/cloud/stream/app/log-sink-kafka/2.1.1.RELEASE/log-sink-kafka-2.1.1.RELEASE.jar
+```
+
+### Custom Processor
+
+For the data transformation between the source and sink steps, we will highlight a custom processor application and use that as a base to demonstrate different programming models.
+
+**Code:**
+
+<!--TABS-->
+<!--Message Channels-->
+
+```java
+@EnableBinding(Processor.class)
+public class SimpleStreamSampleProcessor {
+
+	@StreamListener(Processor.INPUT)
+	@SendTo(Processor.OUTPUT)
+	public String messenger(String data) {
+		return "Hello: " + data + "!";
+	}
+}
+```
+
+<!--Functional-->
+
+```java
+@EnableBinding(Processor.class)
+public class FunctionStreamSampleProcessor {
+
+	@Bean
+	public Function<String, String> messenger() {
+		return data -> "Hello: " + data + "!";
+	}
+}
+```
+
+<!--Kafka Streams-->
+
+```java
+@EnableBinding(KafkaStreamsProcessor.class)
+public class KafkaStreamsSampleProcessor {
+
+	@StreamListener("input")
+	@SendTo("output")
+	public KStream<String, String> messenger(KStream<String, String> data) {
+		return data.map((k, v) -> new KeyValue<>(null, "Hello: " + v + "!"));
+	}
+}
+```
+
+<!--Reactive-->
+
+```java
+@EnableBinding(Processor.class)
+public class ReactiveStreamSampleProcessor {
+
+	@StreamListener(Processor.INPUT)
+	@SendTo(Processor.OUTPUT)
+	public Flux<String> messenger(Flux<String> data) {
+		return data.map(incoming -> "Hello: " + incoming + "!");
+	}
+}
```

Comment:
The old reactive model and the functional model are converging. The functional model under the hood uses Reactor components to do the transformation. We can express the `Flux` transformations in the context of a function bean now. Wonder, we have to collapse these two or wait it out and see how they all shape up in the in the coming releases of Spring Cloud Stream?

Reply:
It makes sense not to list it since it is already soft deprecated. Done.